Add Numeric#sign: T as a replacement of Numeric#signum: Int

eed3si9n opened the pull request:
Fixes scala/bug#11386
See also https://github.com/scala/scala/pull/7748

`Numeric#signum` currently returns Int, which does not handle NaN as reported in scala/bug#11386. Current behavior is inconsistent with Java's `java.lang.Math.signum(Double)`, which returns NaN for Nan, zero for zero, 1.0 for positive, -1.0 for negative. Interestingly, this is also inconsistent with `scala.math.signum(Double)`, which forwards to `java.lang.Math.signum(Double)`!

This changes `Numeric#signum` to return T. For `Int`, `Long`, `Float`, and `Double`, `Numeric#signum` will forward to `scala.math.signum(...)` of the corresponding types. For `Short`, `Byte`, and `Char` it will use `scala.math.signum(Int)` similar to the way `max` is implemented in `RichShort`.

```scala
Welcome to Scala 2.13.0-20190313-035618-c447d5a (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_171).
Type in expressions for evaluation. Or try :help.

scala> Double.NaN.signum
                  ^
       warning: method signum in class RichDouble is deprecated (since 2.13.0): signum does not handle -0.0 or Double.NaN; use `sign` method instead
res0: Int = 0

scala> Double.NaN.sign
res1: Double = NaN

scala> scala.math.signum(Double.NaN)
res2: Double = NaN
```

eed3si9n committed:
change Numeric#signum: Int to return T

Fixes scala/bug#11386

`Numeric#signum` currently returns Int, which does not handle NaN as reported in scala/bug#11386. Current behavior is inconsistent with Java's `java.lang.Math.signum(Double)`, which returns NaN for Nan, zero for zero, 1.0 for positive, -1.0 for negative. Interestingly, this is also inconsistent with `scala.math.signum(Double)`, which forwards to `java.lang.Math.signum(Double)`!

This changes `Numeric#signum` to return T. For `Int`, `Long`, `Float`, and `Double`, `Numeric#signum` will forward to `scala.math.signum(...)` of the corresponding types. For `Short`, `Byte`, and `Char` it will use `scala.math.signum(Int)` similar to the way `max` is implemented in `RichShort`.

diesalbla commented on src/library/scala/math/Numeric.scala:
```diff
@@ -80,6 +81,8 @@ object Numeric {
     def toLong(x: Short): Long = x.toLong
     def toFloat(x: Short): Float = x.toFloat
     def toDouble(x: Short): Double = x.toDouble
+    // delegate to math.signum(Int)
```
IMHO, this comment does not add anything to what can be read from the code.

eed3si9n replied:
At the time it felt somewhat notable that it's using `Int`'s facility, but you're right that adding this comment doesn't add much.

eed3si9n replied:
Addressed in 430c7acb33d41ea95360f613fe9794387fd488b8

diesalbla commented on src/library/scala/math/Numeric.scala:
```diff
@@ -96,6 +99,8 @@ object Numeric {
     def toLong(x: Byte): Long = x.toLong
     def toFloat(x: Byte): Float = x.toFloat
     def toDouble(x: Byte): Double = x.toDouble
+    // delegate to math.signum(Int)
```
Likewise.

diesalbla commented on src/library/scala/math/Numeric.scala:
```diff
@@ -112,6 +117,8 @@ object Numeric {
     def toLong(x: Char): Long = x.toLong
     def toFloat(x: Char): Float = x.toFloat
     def toDouble(x: Char): Double = x.toDouble
+    // delegate to math.signum(Int)
```
Same here.

diesalbla commented on test/junit/scala/math/DoubleTest.scala:
```diff
@@ -0,0 +1,16 @@
+class DoubleTest {
+
+  /* Test for scala/bug#11386 */
+  @Test
```
May I suggest testing also plus/minus infinities and zeroes?

eed3si9n replied:
Done in 430c7acb33d41ea95360f613fe9794387fd488b8

eed3si9n committed:
Add more test for Double#signum

szeiger commented:
The fallout looks relatively minor, which is good. I'd prefer a non-breaking change though. Can we give the method a different name (and possibly deprecate the old one) as suggested in https://github.com/scala/scala/pull/7748#issuecomment-471227858?

eed3si9n commented:
@szeiger 👍 Adding a new method sounds more prudent given the numeric widening.

eed3si9n committed:
Add Numeric#sign: T method

For backward compatibility, add the new behavior as Numeric#sign: T, and deprecate Numeric#signum.

szeiger commented:
Quoting @Ichoran (on the deprecated method):

> I guess for correctness one would want to deprecate it. But then again, having floating-point numbers abstracted into the interface at all is kind of a correctness bug. So I'm not sure.
> 
> My instinct is to leave it alone and document the difference carefully. I don't see a compelling reason to muck up people's already-working code.

I'm inclined to agree with "not sure" and, after looking at the latest changes, also with not deprecating. We could still use the old `signum` for stuff like `NumericRange`. It's probably a bit faster and this code can't deal with abnormal values like NaN anyway.

eed3si9n commented:
`RichDouble#signum` not matching up with `scala.math.signum(Double)` should be treated as a bug, right?

Ichoran commented:
@eed3si9n - Yes, I'd consider that a bug, but tweaking implicits to get the correct behavior for something that gets the functionality from `Numeric` wasn't working out very well due to the relatively clumsy options for adjusting the priority of implicits.  We can always break that link later (i.e. `3.0.signum` doesn't have the same return type as `(3.0: Numeric[Double]).signum`.

eed3si9n commented:
ok. What do you guys think about the following plan:
1. Don't make any changes `Numeric[A]` typeclass at all. Leave things be.
2. Add new `sign` method to `Rich*` (`RichShort`, `RichInt`, `RichDouble`) where `#sign` returns `Int` for everyone except `RichDouble` and `RichFloat` (they'll return `Double` and `Float` respectively)
3. Deprecate `signum` method only on `RichDouble` and `RichFloat`, and tell users to migrate to `sign`
4. In 2.14, get `signum` out of `ScalaNumberProxy` so we can remove `RichDouble#signum`

szeiger commented:
I'd start by adding the new method to `Numeric`. That's the one place where you really need it so you can abstract over different numeric types and still get the correct behavior for floating-point types.

Anything else (`RichDouble`, `scala.math`, `java.lang.Math`) is used for a specific numeric type. In this case you know what type you have and which signum function you need so can just pick the right one. It would still be nice to keep them consistent with `Numeric`.

eed3si9n commented:
@szeiger It depends what we think the right behavior is. Suppose we want `signum` to return `Int` for most numeric types, except for `Double` and `Float`, then that would be difficult to express that either as `Numeric[T]` or `ScalaNumberProxy[T]` which is a parent of `RichDouble`. If we want genericity, returning `T` probably makes sense, and that's what I've done with `sign: T` in b0c2beb. Would you be ok if I just removed deprecation of signum?

The plan I outlined in https://github.com/scala/scala/pull/7861#issuecomment-473081273 goes the opposite direction, and give up on genericity but limit the fix to `RichDouble`, which in practical situation probably arise more frequently than `Numeric`.

Ichoran commented:
@szeiger - The problem is that it's not terribly clear what the correct behavior _is_ for floating-point types in a generic context.  This isn't limited to signum; everything is a mess when NaN gets involved.

Working generically, you expect various things to be true, like either `a < b` or `a >= b`.  Well, not so with floating-point!  You expect `abs(x)*signum(x) == x`, which _can_ be true with `Int` signum, but isn't in our (or `java.lang.Math.signum(x)`'s) implementation.  You expect that `abs(x) * (-abs(x))` is negative, but not if `x` is `NaN`.  When working generically, you can't even test any of these things because you can't talk about `NaN`.

So I'm happy with the plan to de-generify `sign` now.  I'm also happy with it being generic.  I just don't think it matters very much.  The critical thing is to have something on `RichDouble` that matches `math.signum`.

Ichoran reviewed (approved):
The implementation looks good to me.  There are a number of sort of tricky issues to think through regarding Double, but I think it's all okay.

I have doubts about deprecating `signum`.  It pretty much works, except for `Double` and `Float` where everything, not just signum, is busted when you admit `NaN`.  And most people are going to be familiar with Java, and will want `signum`.  But if we're going to, this looks like a reasonable way to do it (and avoid the signum-of-NaN bug which is a bug).

IEEE754 doesn't suggest that you implement `signum`.  Instead, it recommends `copysign`.  Maybe we should implement that instead/as well.
